File: src/main/java/com/manabreakstudios/discourse/ui/core/editor/SlotComponent.java
package com.manabreakstudios.discourse.ui.core.editor;

import com.manabreakstudios.discourse.ui.Theme;
import lombok.Getter;
import lombok.RequiredArgsConstructor;

import javax.swing.*;
import java.awt.*;

import static com.manabreakstudios.discourse.ui.core.utils.SwingUtils.drawCircle;

@RequiredArgsConstructor
public class SlotComponent extends JPanel {

    @Getter
    private final SlotBinding slotBinding;

    public Slot getSlot() {
        return slotBinding.getSlot();
    }

    @Override
    protected void paintComponent(Graphics g) {
        Graphics2D g2d = (Graphics2D) g;
        g2d.setRenderingHint(RenderingHints.KEY_ANTIALIASING, RenderingHints.VALUE_ANTIALIAS_ON);

        int slotSize = Theme.theme().getSlotSize();
        int slotBorderWidth = Theme.theme().getSlotBorderWidth();

        g2d.setColor(Theme.theme().getNodeBorderColor());
        drawCircle(g2d, 0, 0, slotSize);

        g2d.setColor(Theme.theme().getSlotColor());
        drawCircle(g2d, slotBorderWidth, slotBorderWidth, slotSize - 2 * slotBorderWidth);
    }
}

File: node-editor/src/com/manabreak/node_editor/SlotComponent.kt
package com.manabreak.node_editor

import com.manabreak.node_editor.Renderer.createSlotImage
import com.manabreak.node_editor.Theme.Companion.theme
import com.manabreak.node_editor.model.Slot
import java.awt.Graphics
import java.awt.Graphics2D
import java.awt.RenderingHints
import java.awt.image.BufferedImage
import javax.swing.JPanel

class SlotComponent constructor(val slot: Slot) : JPanel() {

    var state = State.NORMAL

    override fun paintComponent(g: Graphics) {
        val g2d = g as Graphics2D
        g2d.setRenderingHint(RenderingHints.KEY_ANTIALIASING, RenderingHints.VALUE_ANTIALIAS_ON)
        g2d.drawImage(state.image, 0, 0, null)
    }

    enum class State constructor(image: BufferedImage) {
        NORMAL(createSlotImage(theme.nodeBorderColor, theme.slotDefaultColor, theme.slotBorderWidth)),
        ACCEPT(createSlotImage(theme.nodeBorderColor, theme.slotAcceptColor, theme.slotBorderWidth)),
        DECLINE(createSlotImage(theme.nodeBorderColor, theme.slotDeclineColor, theme.slotBorderWidth));

        val image: BufferedImage

        init {
            this.image = image
        }
    }
}
